provider/ec2/iam.go

```diff
@@ -0,0 +1,176 @@
+// Copyright 2021 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package ec2
+
+import (
+	stdcontext "context"
+	stderrors "errors"
+	"fmt"
+	"time"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/service/ec2"
+	ec2types "github.com/aws/aws-sdk-go-v2/service/ec2/types"
+	"github.com/aws/aws-sdk-go-v2/service/iam"
+	iamtypes "github.com/aws/aws-sdk-go-v2/service/iam/types"
+	"github.com/juju/clock"
+	"github.com/juju/errors"
+	"github.com/juju/retry"
+
+	"github.com/juju/juju/core/instance"
+	"github.com/juju/juju/core/status"
+	"github.com/juju/juju/environs"
+	"github.com/juju/juju/environs/cloudspec"
+	"github.com/juju/juju/environs/context"
+	"github.com/juju/juju/environs/instances"
+)
+
+// instanceProfileClient is a subset interface of the ec2 client for attaching
+// Instance Profiles to ec2 machines.
+type instanceProfileClient interface {
+	AssociateIamInstanceProfile(stdcontext.Context, *ec2.AssociateIamInstanceProfileInput, ...func(*ec2.Options)) (*ec2.AssociateIamInstanceProfileOutput, error)
+}
+
+// IAMClient is a subset interface of the AWS IAM client. This interface aims
+// to define the small set of what Juju's needs from the larger client.
+type IAMClient interface {
+	CreateInstanceProfile(stdcontext.Context, *iam.CreateInstanceProfileInput, ...func(*iam.Options)) (*iam.CreateInstanceProfileOutput, error)
+	GetInstanceProfile(stdcontext.Context, *iam.GetInstanceProfileInput, ...func(*iam.Options)) (*iam.GetInstanceProfileOutput, error)
+}
+
+// IAMClientFunc defines a type that can generate an AWS IAMClient from a
+// provided cloudspec.
+type IAMClientFunc = func(stdcontext.Context, cloudspec.CloudSpec, ...ClientOption) (IAMClient, error)
+
+const (
+	setProfileDelay      = time.Second * 5
+	setProfileMaxAttempt = 5
+)
+
+// iamClientFunc implements the IAMClientFunc type and is used internally by
+// Juju for creating an IAM client.
+func iamClientFunc(
+	ctx stdcontext.Context,
+	spec cloudspec.CloudSpec,
+	clientOptions ...ClientOption,
+) (IAMClient, error) {
+	cfg, err := configFromCloudSpec(ctx, spec, clientOptions...)
+	if err != nil {
+		return nil, errors.Annotate(err, "building aws config from cloudspec")
+	}
+	return iam.NewFromConfig(cfg), nil
+}
+
+// ensureControllerInstanceProfile ensures that a controller Instance Profile
+// has been created for the supplied controller name in the specified AWS cloud.
+func ensureControllerInstanceProfile(
```

Comment:
Can this just be called `ensureInstanceProfile`
There's nothing controller specific here apart from the profile name and we can just pass that in when called.